### DISCUSSION TITLE
官方示例的随机数到底应该怎么用

### DISCUSSION THREAD

**Discussion by passer-byzhang** [NonceGeek/Move-Camp#264] [Q&A] [answered]
官方库里看到了两个
https://github.com/MystenLabs/sui/blob/main/sui_programmability/examples/games/sources/drand_based_lottery.move
与
https://github.com/MystenLabs/sui/blob/main/sui_programmability/examples/games/sources/randomness_based_lottery.move

里边每一轮传入字节和验证字节没看懂，是mpc吗，这个最后是否可以用来支撑财产相关的业务？

**Comment by qiwihui** [comment, marked as answer]
这两个lottery 是使用的是不同的的随机过程：
1. drand_based_lottery: 
    - 使用的是 drand 分布式随机信标，这个随机信标去中心化、可抵御偏差、公开可验证及可靠，随机性信标每 30 秒更新一次 （round），可以根据 genesis 时间计算当前 round 数。 具体可参见 https://drand.love/。
    - 游戏建议的输入是未来五分钟的 round 数字，只要有人在 round - 1 之前关闭（close）对应的 lottery，就可以认为这个lottery 是随机的，游戏为验证 round - 2 的签名。
    - 每个 round 签名用的是椭圆曲线 BLS12381 进行验证，可以看代码 https://github.com/MystenLabs/sui/blob/main/sui_programmability/examples/games/sources/drand_lib.move#L49 ，通过 `prev_drand_sig` 和 `round` 生成 digest 来验证 `drand_sig `。 验证后用 drand_sig 生成随机性（derive_randomness，sha256）作为选择赢家（safe_selection）的种子，代码参见 https://github.com/MystenLabs/sui/blob/main/sui_programmability/examples/games/sources/drand_based_lottery.move#L108-L115。
2. randomness_based_lottery：
    - 使用的是区块的 BLS 签名作为验证：创建时，Randomness 对象对应包含 epoch 和对象ID。创建交易结束后，任何人可以从 validators 那边拿到 `"randomness":epoch:id` 这个消息的 BLS 签名。下一个交易通过这个签名修改 Randomness 中的 value，value 值为 BLS 签名的 sha256 哈希，value 的值作为随机种子进行选择赢家（safe_selection），代码可见： https://github.com/MystenLabs/sui/blob/main/crates/sui-framework/sources/crypto/randomness.move#L77-L83。
    - 验证签名的算法是原生的 BLS 验证： `native_tbls_verify_signature`。
    - Randomness 对象不能在*相同*交易中创建和使用，因为它可能允许验证者在看到它们所依赖的随机性后决定是否创建和使用这些对象。

基于 randomness 的还不能做为生产业务合约使用，仅能用于测试，因为其随机性的基础不安全，官方在推进中。 Drand 的可以。

**Reply by passer-byzhang** [reply]
感谢您，讲的很清楚！
